### PR TITLE
[wip] Parameterized CF template

### PR DESCRIPTION
**Pull request by sethfitz**
Creates a VPC, etc. to run within.

This approach to templating doesn't work, interpolated values look like `THING=<troposphere.Ref object at 0x10c5c3e10>`.  @hectcastro any ideas on how to do this better?

Remaining:
- [x] SQS queue
- [x] S3 buckets
- [x] S3 bucket policies
- [x] IAM role
- [x] IAM policy
- [x] associate IAM role w/ instance

Refs #5
Refs #6

### CODE REVIEW

File: cloud-config/oam-server-api.yml [hunk 1]
```diff
@@ -26,5 +63,13 @@ coreos:
...
-        ExecStart=/usr/bin/docker run --rm --name oam-server-publisher quay.io/hotosm/oam-server-publisher:v0.1.0
+        ExecStart=/usr/bin/docker run \
+          --rm \
+          --name oam-server-publisher \
+          -e 'AWS_DEFAULT_REGION={aws_region}' \
+          -e 'OAM_STATUS_SQS_QUEUE_URL={status_sqs_queue_url}' \
+          -e 'OAM_API_TOKEN={oam_api_token}' \
+          -e 'OAM_STATUS_BUCKET={status_bucket}' \
+          -e 'OAM_STATUS_PREFIX={status_prefix}' \
+          quay.io/hotosm/oam-server-api:{publisher_version}
```
Comment (hectcastro) [inline cloud-config/oam-server-api.yml:74]: Not sure if the responsibilities of the services has changed, but you have this container named as `oam-server-publisher`, but are using the `oam-server-api` container image.

File: cloud-config/oam-server-api.yml [hunk 2]
```diff
@@ -4,6 +4,21 @@ coreos:
...
   units:
+    - name: cfn-signal.service
+      command: start
+      content: |
+        [Unit]
+        Description=CloudFormation signaler
+        After=docker.service
+
+        [Service]
+        ExecStart=/usr/bin/docker run \
```
Comment (hectcastro) [inline cloud-config/oam-server-api.yml:15]: Should this be a `oneshot` service?

File: cfn/tiler_api_stack.py [hunk 3]
```diff
@@ -11,242 +14,379 @@
...
+# convert to JSON representations
+refs = dict([(k, '|||' + json.dumps(v.JSONrepr()) + '|||') for (k, v) in refs.items()])
+
+# interpolate and split into an array
+user_data = read_file('cloud-config/oam-server-api.yml').format(**refs).split('|||')
+
+# replace stringified refs with objects
+user_data = map(lambda x: json.loads(x) if x.startswith('{"Ref": "') else x, user_data)
+
+instance_profile = t.add_resource(iam.InstanceProfile(
```
Comment (hectcastro) [inline cfn/tiler_api_stack.py:343]: You needed to create this as a resource to reference it?

File: cfn/tiler_api_stack.py [hunk 4]
```diff
@@ -11,242 +14,379 @@
...
+    auth_token_bucket=Ref(auth_token_bucket_param),
+    auth_token_key=Ref(auth_token_key_param),
+    status_bucket=Ref(status_bucket_param),
+    status_prefix=Ref(status_prefix_param),
+    status_sqs_queue_url=Ref(status_sqs_queue_url_param),
+    oam_api_token=Ref(oam_api_token_param),
+)
+
+# convert to JSON representations
+refs = dict([(k, '|||' + json.dumps(v.JSONrepr()) + '|||') for (k, v) in refs.items()])
```
Comment (hectcastro) [inline cfn/tiler_api_stack.py:335]: I think that I would have ended up doing something similar given the number of variables being passed in. Only thing additional I would have offered would have been to use something like Jinja2 to do the templating, but this seems to be working fine.

File: Makefile [hunk 5]
```diff
@@ -6,7 +6,7 @@ all: $(CFN_JSON_FILES)
 clean:
 	rm -f cfn/*.json
 
-cfn/%.json: cfn/%.py
+cfn/%.json: cfn/%.py cloud-config/*.yml
```
Comment (hectcastro) [inline Makefile:9]: I should have had this in there from the start.

File: cfn/tiler_api_stack.py (hunk 3 above)
Comment (sethfitz) [inline cfn/tiler_api_stack.py:343]: I think so (didn't try without it).  My reference source was https://github.com/cloudtools/troposphere/blob/master/examples/IAM_Roles_and_InstanceProfiles.py

File: cloud-config/oam-server-api.yml (hunk 2 above)
Comment (sethfitz) [inline cloud-config/oam-server-api.yml:15]: Yes.  Is this syntax right?  (I'm more familiar with `upstart`.)

```
[Unit]
Description=CloudFormation signaler
After=docker.service

[Service]
Type=oneshot
ExecStart=/usr/bin/docker run \
  --rm \
  quay.io/mojodna/awscli \
  cfn-signal \
  --stack {aws_stack_name} \
  --resource WebServerInstance \
  --region {aws_region}
```

File: cloud-config/oam-server-api.yml (hunk 1 above)
Comment (sethfitz) [inline cloud-config/oam-server-api.yml:74]: Good eye.  I fixed this in a later commit, but not before it caused errors ;-)

File: cloud-config/oam-server-api.yml (hunk 2 above)
Comment (hectcastro) [inline cloud-config/oam-server-api.yml:15]: Yep.

File: cfn/tiler_api_stack.py (hunk 3 above)
Comment (hectcastro) [inline cfn/tiler_api_stack.py:343]: Hm. If the role already exists, I was able to just pass the string name through:

``` python
app_server_instance_profile_param = t.add_parameter(Parameter(
    'AppServerInstanceProfile', Type='String',
    Default='AppServerInstanceProfile',
    Description='AWS::IAM::Role name for the application servers'
))

...

app_server_launch_config = t.add_resource(asg.LaunchConfiguration(
    'lcAppServer',
    ImageId=Ref(app_server_ami_param),
    IamInstanceProfile=Ref(app_server_instance_profile_param),
    InstanceType=Ref(app_server_instance_type_param),
    KeyName=Ref(keyname_param),
    SecurityGroups=[Ref(app_server_security_group)]
))
```

**Pull request merged**